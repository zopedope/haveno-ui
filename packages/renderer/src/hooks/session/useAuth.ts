// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useQuery } from "react-query";
import { QueryKeys } from "@constants/query-keys";
import { validateSession } from "@utils/session";

export function useAuth() {
  return useQuery<boolean, Error>(
    QueryKeys.AuthSession,
    async () => {
      if (await validateSession()) {
        return true;
      }
      return false;
    },
    {
      staleTime: 60000,
      retry: false,
    }
  );
}
