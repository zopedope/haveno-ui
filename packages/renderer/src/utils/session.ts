// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

const SESSION_KEY = "AUTH_TOKEN";

export async function createSession(authToken: string) {
  window.sessionStorage.setItem(SESSION_KEY, authToken);
}

export async function validateSession(): Promise<boolean> {
  const token = window.sessionStorage.getItem(SESSION_KEY);
  if (!token) {
    return false;
  }
  return window.electronStore.verifyAuthToken(token);
}

export async function deleteSession() {
  window.sessionStorage.removeItem(SESSION_KEY);
}
