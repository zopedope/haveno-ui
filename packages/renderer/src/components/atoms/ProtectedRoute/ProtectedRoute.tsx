// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useEffect } from "react";
import { useNavigate } from "react-router-dom";
import type { ReactNode } from "react";
import { useAuth } from "@hooks/session/useAuth";
import { deleteSession } from "@utils/session";
import { ROUTES } from "@constants/routes";

export function ProtectedRoute({ children }: { children: ReactNode }) {
  const { data: isAuthed, isLoading, isSuccess } = useAuth();
  const navigate = useNavigate();

  useEffect(() => {
    if (isLoading) {
      return;
    }
    if (!isAuthed || !isSuccess) {
      deleteSession();
      navigate(ROUTES.Login);
    }
  }, [isLoading, isAuthed, isSuccess]);

  return isSuccess && isAuthed ? <>{children}</> : null;
}
