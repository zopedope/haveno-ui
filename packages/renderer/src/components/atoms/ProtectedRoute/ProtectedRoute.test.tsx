// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

const navigateSpy = vi.fn();
const deleteSessionSpy = vi.fn();

import { beforeAll, describe, expect, it, vi, afterEach } from "vitest";
import { render, screen } from "@testing-library/react";
import type { SpyInstanceFn } from "vitest";
import { ProtectedRoute } from "./ProtectedRoute";
import { useAuth } from "@hooks/session/useAuth";
import { ROUTES } from "@constants/routes";

describe("atoms::ProtectedRoute", () => {
  beforeAll(() => {
    vi.mock("react-router-dom", () => ({
      useNavigate: () => navigateSpy,
    }));

    vi.mock("@utils/session", () => ({
      deleteSession: deleteSessionSpy,
    }));

    vi.mock("@hooks/session/useAuth", () => ({
      useAuth: vi.fn(() => ({
        isLoading: true,
        isSuccess: false,
        data: undefined,
      })),
    }));
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it("renders children if auth session exists", () => {
    (useAuth as SpyInstanceFn).mockReturnValue({
      isLoading: false,
      isSuccess: true,
      data: true,
    });
    const { asFragment, unmount } = render(
      <ProtectedRoute>
        <div>Protected content</div>
      </ProtectedRoute>
    );
    expect(screen.queryByText("Protected content")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("skips rendering children and redirects to login if no auth session exists", () => {
    (useAuth as SpyInstanceFn).mockReturnValueOnce({
      isLoading: false,
      isSuccess: true,
      data: false,
    });

    const { asFragment, unmount } = render(
      <ProtectedRoute>
        <div>Protected content</div>
      </ProtectedRoute>
    );
    expect(screen.queryByText("Protected content")).toBeNull();
    expect(deleteSessionSpy).toHaveBeenCalledOnce();
    expect(navigateSpy).toHaveBeenCalledOnce();
    expect(navigateSpy).toHaveBeenCalledWith(ROUTES.Login);
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("skips rendering children and redirects to login if auth session can't be retrieved", () => {
    (useAuth as SpyInstanceFn).mockReturnValueOnce({
      isLoading: false,
      isSuccess: false,
      data: true,
    });

    const { asFragment, unmount } = render(
      <ProtectedRoute>
        <div>Protected content</div>
      </ProtectedRoute>
    );
    expect(screen.queryByText("Protected content")).toBeNull();
    expect(deleteSessionSpy).toHaveBeenCalledOnce();
    expect(navigateSpy).toHaveBeenCalledOnce();
    expect(navigateSpy).toHaveBeenCalledWith(ROUTES.Login);
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });
});
