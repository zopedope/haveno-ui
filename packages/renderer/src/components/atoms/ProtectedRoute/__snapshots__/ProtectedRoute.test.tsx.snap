// Vitest Snapshot v1

exports[`atoms::ProtectedRoute > renders children if auth session exists 1`] = `
<DocumentFragment>
  <div>
    Protected content
  </div>
</DocumentFragment>
`;

exports[`atoms::ProtectedRoute > skips rendering children and redirects to login if auth session can't be retrieved 1`] = `<DocumentFragment />`;

exports[`atoms::ProtectedRoute > skips rendering children and redirects to login if no auth session exists 1`] = `<DocumentFragment />`;
